wit/request-reply.wit

```diff
@@ -0,0 +1,43 @@
+/// The request-reply interface allows a guest to send a message and await a response. This
+/// interface is considered optional as not all message services support the concept of
+/// request/reply. However, request/reply is a very common pattern in messaging and as such, we have
+/// included it as a core interface.
+interface request-reply {
+    use types.{client, message, error};
+
+    /// Options for a request/reply operation. This is a resource to allow for future expansion of
+    /// options.
+    resource request-options {
+        /// Creates a new request options resource with no options set.
+        constructor();
+
+        /// The maximum amount of time to wait for a response. If the timeout value is not set, then
+        /// the request/reply operation will block until a message is received in response.
+        set-timeout-ms: func(timeout-ms: u32);
+
+        /// The maximum number of replies to expect before returning. This only applies to
+        /// `request-multi` and is ignored otherwise. If the number of replies is not set and
+        /// timeout isn't set, then the operation will block until a message is received in
+        /// response (essentially the same behavior as `request`).
+        set-expected-replies: func(expected-replies: u32);
+    }
+
+    /// Performs a blocking request/reply operation with an optional set of request options. This
+    /// returns only the first reply received or a timeout . If more than one reply is expected, then the
+    /// `request-multi` function should be used instead.
+    request: func(c: borrow<client>, msg: message, opts: option<request-options>) -> result<message, error>;
+
+    /// Performs a blocking request/reply operation with an optional set of request options. This
+    /// returns all replies received up until timeout or the configured set of expected replies. It
+    /// is recommended to use a `request-options` with the timeout set to ensure that the operation
+    /// does not block indefinitely. Unlike request, this function should not return an error on
+    /// timeout and should instead return all of the replies received up to that point. This is to
+    /// faciliate use in scatter/gather operations where the number of expected replies is not
+    /// known.
+    request-multi: func(c: borrow<client>, msg: message, opts: option<request-options>) -> result<list<message>, error>;
```

Comment:
```suggestion
    request-multi: func(c: borrow<client>, msg: message, opts: option<request-multi-options>) -> result<list<message>, error>;
```

I would love different data structures here

Reply:
Currently, `request-options` only has two functions:
- `set-timeout-ms`, and
- `set-expected-replies`.

I understand `set-expected-replies` isn't used by `request` and only `request-multi`, but I feel like adding multiple structures could just add some confusion.

Do you feel like there are fields missing that would warrant the separation of `request-options` and `request-multi-options`? Otherwise, I'd opt to keeping them the same 👍

Reply:
My concern is not "now" but "tomorrow." The options have two different use cases, and a copy-and-paste today could avoid a major breaking change in the future when configurations collide.

Cohesion around the use-case will avoid problematic global configuration options that the user needs to learn how to configure properly per use case.

Reply:
At the same time, copy-and-paste here would mean that a host would have to copy-and-paste their implementation for, say, `set-timeout-ms`. This could lead to divergence in implementations that should be the same and hurt maintainability. That said, I understand what you're saying - maybe there's a middleground somewhere here...

Reply:
Yeah, there's no question about your concern here.

Overall, let the implementor decide how much code-reusing they want to do and how the code path may look.

I am purely focusing on the specification detail.